### PR TITLE
RATIS-2036 Avoid trigger snapshot when removing raftGroup

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**
see [jira](https://issues.apache.org/jira/browse/RATIS-2036)

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java [hunk 1]
```diff
@@ -327,12 +332,30 @@ private boolean shouldTakeSnapshot() {
...
+   * true false false => true
+   * false true true => true
+   * false true false =>  false
+   * false false true => false
+   * false false false => false
+   * @return result
+   */
+  private boolean shouldTakeSnapshotAccordingToConfAndStatus() {
+    return (triggerSnapshotWhenStopEnabled || isRemoving)
+        && (triggerSnapshotWhenRemoveEnabled || !isRemoving);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java:356]: Let's call it `shouldTakeSnapshotAtStop`.  We may simply the logic as below:
```java
  private boolean shouldTakeSnapshotAtStop() {
    if (isRemoving) {
      return triggerSnapshotWhenRemoveEnabled;
    } else {
      return triggerSnapshotWhenStopEnabled;
    }
  }
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java [hunk 2]
```diff
@@ -341,6 +364,10 @@ private void notifyAppliedIndex(long index) {
     appliedIndexConsumer.accept(index);
   }
 
+  void setRemoving() {
+    this.isRemoving = true;
+  }
+
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java:370]: Let's move it to right below `shouldTakeSnapshotAtStop()`.

**Review by szetszwo — COMMENTED** [review]
@OneSizeFitsQuorum , thanks for working on this!  Then change looks good.  Just two minor comments inlined.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java (hunk 1 above)
Comment (OneSizeFitsQuorum) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java:356]: The codes looks much clearer!

File: ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java (hunk 2 above)
Comment (OneSizeFitsQuorum) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java]: OK

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**